sklearn/gaussian_process/gpr.py

```diff
@@ -209,12 +217,17 @@ def obj_func(theta, eval_gradient=True):
                         "Multiple optimizer restarts (n_restarts_optimizer>0) "
                         "requires that all bounds are finite.")
                 bounds = self.kernel_.bounds
-                for iteration in range(self.n_restarts_optimizer):
+
+                def optima_iterations():
                     theta_initial = \
                         self.rng.uniform(bounds[:, 0], bounds[:, 1])
                     optima.append(
                         self._constrained_optimization(obj_func, theta_initial,
                                                        bounds))
+                Parallel(n_jobs=self.n_jobs)(delayed(optima_iterations,
+                                                     check_pickle=False)()
```

Comment:
To illustrate what I mean:

```py
from joblib import Parallel, delayed

import numpy as np


def func(rng):
    return rng.randn(2)

rng = np.random.RandomState(0)

n_jobs=2
result = Parallel(n_jobs=n_jobs)(delayed(func)(rng) for x in [1, 2, 3])
print(result)
```

with n_jobs=1:
```
[array([ 1.76405235,  0.40015721]), array([ 0.97873798,  2.2408932 ]), array([ 1.86755799, -0.97727788])]
```

with n_jobs=2
```
[array([ 1.76405235,  0.40015721]), array([ 1.76405235,  0.40015721]), array([ 1.76405235,  0.40015721])]
```

What happens in the n_jobs=2 case is that each subprocess gets a copy of rng so the random numbers are the same for each value of x.